File: dynamic-property-zk/src/main/java/ru/fix/dynamic/property/zk/ZkDynamicPropertySource.java
package ru.fix.dynamic.property.zk;

import org.apache.curator.framework.CuratorFramework;
import org.apache.curator.framework.CuratorFrameworkFactory;
import org.apache.curator.framework.imps.CuratorFrameworkState;
import org.apache.curator.framework.recipes.cache.ChildData;
import org.apache.curator.framework.recipes.cache.TreeCache;
import org.apache.curator.framework.recipes.cache.TreeCacheEvent;
import org.apache.curator.retry.ExponentialBackoffRetry;
import org.apache.zookeeper.KeeperException;
import org.apache.zookeeper.data.Stat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ru.fix.dynamic.property.api.DynamicPropertyListener;
import ru.fix.dynamic.property.api.DynamicPropertySource;
import ru.fix.dynamic.property.api.marshaller.DynamicPropertyMarshaller;

import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.Function;

/**
 * Implementation of {@link DynamicPropertyListener} which uses {@link TreeCache} inside and
 * provides subscriptions to property change events
 */
public class ZkDynamicPropertySource implements DynamicPropertySource {

    private static final Logger log = LoggerFactory.getLogger(ZkDynamicPropertySource.class);

    private static final int UPSERT_PROPERTY_RETRY_COUNT = 10;

    private final String configLocation;
    private CuratorFramework curatorFramework;
    private final DynamicPropertyMarshaller marshaller;

    private Map<String, Collection<DynamicPropertyListener<String>>> listeners = new ConcurrentHashMap<>();

    private TreeCache treeCache;

    public ZkDynamicPropertySource(String zookeeperQuorum, String configLocation, DynamicPropertyMarshaller marshaller) throws Exception {
        this(
                CuratorFrameworkFactory.newClient(
                        zookeeperQuorum,
                        new ExponentialBackoffRetry(1000, 10)
                ),
                configLocation,
                marshaller
        );
    }

    /**
     * @param curatorFramework Ready to use curator framework
     * @param configLocation   Root path where ZkDynamicPropertySource will store properties. E.g.
     *                         '/cpapsm/config/SWS'
     */
    public ZkDynamicPropertySource(CuratorFramework curatorFramework, String configLocation, DynamicPropertyMarshaller marshaller) throws Exception {
        this.curatorFramework = curatorFramework;
        this.configLocation = configLocation;
        this.marshaller = marshaller;
        init();
    }

    private void init() throws Exception {
        if (curatorFramework.getState().equals(CuratorFrameworkState.LATENT)) {
            curatorFramework.start();
        }

        treeCache = new TreeCache(this.curatorFramework, this.configLocation);
        treeCache.getListenable().addListener((currentFramework, treeCacheEvent) -> {
            switch (treeCacheEvent.getType()) {
                case NODE_ADDED:
                case NODE_UPDATED:
                    firePropertyChanged(treeCacheEvent, path -> {
                        try {
                            return new String(currentFramework.getData().forPath(treeCacheEvent.getData().getPath()),
                                    StandardCharsets.UTF_8);
                        } catch (Exception e) {
                            log.error("Zk property updating error", e);
                        }
                        return null;
                    });
                    break;
                case NODE_REMOVED:
                    firePropertyChanged(treeCacheEvent, path -> null);
                    break;
                default:
                    break;
            }
        });
        treeCache.start();
    }

    private void firePropertyChanged(TreeCacheEvent treeCacheEvent, Function<String, String> valueExtractor) {
        String propertyPath = treeCacheEvent.getData().getPath();
        Collection<DynamicPropertyListener<String>> zkPropertyChangeListeners = listeners.get(propertyPath);
        String newValue = valueExtractor.apply(propertyPath);
        log.info("Event type {} for node '{}'. New value is '{}'", treeCacheEvent.getType(), propertyPath, newValue);

        if (zkPropertyChangeListeners != null) {
            zkPropertyChangeListeners.forEach(listener -> {
                try {
                    listener.onPropertyChanged(newValue);
                } catch (Exception e) {
                    log.error("Failed to update property {}", propertyPath, e);
                }

            });
        }
    }

    @Override
    public void upsertProperty(String key, String propVal) throws Exception {
        String propPath = getAbsolutePath(key);
        int iteration = 0;
        do {
            ChildData currentData = treeCache.getCurrentData(propPath);
            byte[] newData = propVal.getBytes(StandardCharsets.UTF_8);
            if (currentData != null) {
                if (!Arrays.equals(currentData.getData(), newData)) {
                    curatorFramework.setData().forPath(propPath, newData);
                }
                break;
            }

            try {
                curatorFramework.create().creatingParentsIfNeeded().forPath(propPath, newData);
            } catch (KeeperException.NodeExistsException e) {
                iteration++;
                log.debug("upserting property '{}'='{}', iteration {}", propPath, propVal, iteration);
                if (iteration < UPSERT_PROPERTY_RETRY_COUNT) {
                    continue;
                }

                throw e;
            }
            break;
        } while (true);
    }

    @Override
    public <T> void putIfAbsent(String key, T propVal) throws Exception {
        String propPath = getAbsolutePath(key);
        ChildData currentData = treeCache.getCurrentData(propPath);
        if (currentData == null) {
            curatorFramework.create()
                    .creatingParentsIfNeeded()
                    .forPath(propPath,
                            marshaller.marshall(propVal).getBytes(StandardCharsets.UTF_8));
        }
    }

    private String getProperty(String key) {
        return getProperty(key, (String) null);
    }

    private String getProperty(String key, String defaulValue) {
        String path = getAbsolutePath(key);
        ChildData currentData = treeCache.getCurrentData(path);
        return currentData == null ? defaulValue : new String(currentData.getData(), StandardCharsets.UTF_8);
    }

    @Override
    public <T> T getProperty(String key, Class<T> type) {
        return getProperty(key, type, null);
    }

    @Override
    public <T> T getProperty(String key, Class<T> type, T defaultValue) {
        String value = getProperty(key);
        if (value != null) {
            return marshaller.unmarshall(value, type);
        }
        return defaultValue;
    }

    /**
     * Works through curator directly to load latest data
     */
    @Override
    public Properties getAllProperties() throws Exception {
        Properties allProperties = new Properties();
        Stat exist = curatorFramework.checkExists().forPath(getAbsolutePath(""));
        if (exist != null) {
            List<String> childs = curatorFramework.getChildren().forPath(getAbsolutePath(""));
            if (!childs.isEmpty()) {
                CountDownLatch latcher = new CountDownLatch(childs.size());
                for (String child : childs) {
                    curatorFramework.getData().watched().inBackground((client, event) -> {
                        allProperties.put(child, new String(event.getData(), StandardCharsets.UTF_8));
                        latcher.countDown();
                    }).forPath(getAbsolutePath(child));
                }
                if (!latcher.await(120, TimeUnit.SECONDS)) {
                    throw new TimeoutException("Failed to extract zk properties data");
                }
            }
        }
        return allProperties;
    }

    @Override
    public void updateProperty(String key, String value) throws Exception {
        String path = getAbsolutePath(key);
        curatorFramework.setData().forPath(path, value.getBytes(StandardCharsets.UTF_8));
    }

    @Override
    public <T> void addPropertyChangeListener(String propertyName, Class<T> type,
                                              DynamicPropertyListener<T> typedListener) {
        addPropertyChangeListener(propertyName, value -> {
            T convertedValue = marshaller.unmarshall(value, type);
            typedListener.onPropertyChanged(convertedValue);
        });
    }

    private void addPropertyChangeListener(String propertyName, DynamicPropertyListener<String> listener) {
        listeners.computeIfAbsent(getAbsolutePath(propertyName), key -> new CopyOnWriteArrayList<>()).add(listener);
    }

    private String getAbsolutePath(String nodeName) {
        boolean nodeNameIsEmpty = null == nodeName || nodeName.isEmpty();
        return configLocation + (nodeNameIsEmpty ? "" : '/' + nodeName);
    }

    @Override
    public void close() {
        treeCache.close();
    }
}

File: dynamic-property-zk/src/main/kotlin/ru/fix/dynamic/property/zk/ZkDynamicPropertySource.kt
package ru.fix.dynamic.property.zk

import org.apache.curator.framework.CuratorFramework
import org.apache.curator.framework.CuratorFrameworkFactory
import org.apache.curator.framework.imps.CuratorFrameworkState
import org.apache.curator.framework.recipes.cache.TreeCache
import org.apache.curator.framework.recipes.cache.TreeCacheEvent
import org.apache.curator.framework.recipes.cache.TreeCacheListener
import org.apache.logging.log4j.kotlin.Logging
import ru.fix.dynamic.property.api.marshaller.DynamicPropertyMarshaller
import ru.fix.dynamic.property.api.source.DynamicPropertySource
import ru.fix.dynamic.property.api.source.OptionalDefaultValue
import ru.fix.dynamic.property.std.source.PropertySourceAccessor
import ru.fix.dynamic.property.std.source.PropertySourcePublisher
import ru.fix.stdlib.reference.ReferenceCleaner
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock


/**
 * Implementation of [DynamicPropertySource] that uses Zookeeper [TreeCache]
 * and provides subscriptions to property change events.
 *
 * During initialization constuctor will block and wait until all properties will be loaded from zookeeper to local cache.
 *
 * Use [CuratorFrameworkFactory] in order to get curatorFramework instance.
 * ```
 * CuratorFrameworkFactory.newClient("zk-host1:port1,zk-host2:port2...", ExponentialBackoffRetry(1000, 10))
 * ```
 *
 * @param curatorFramework Ready to use curator framework
 * @param zookeeperConfigPath Root path where ZkDynamicPropertySource will store properties.
 *                            E.g. '/my-application/config'
 * @param initializationTimeout how log to wait until all property values will be loaded
 */
class ZkDynamicPropertySource(
    private val curatorFramework: CuratorFramework,
    zookeeperConfigPath: String,
    marshaller: DynamicPropertyMarshaller,
    initializationTimeout: Duration
) : DynamicPropertySource {

    companion object : Logging

    private val readTreeAndProcessNotificationLock = ReentrantLock()

    private val propertySourcePublisher = PropertySourcePublisher(
        propertySourceAccessor = object : PropertySourceAccessor {
            override fun accessPropertyUnderLock(propertyName: String, accessor: (String?) -> Unit) {
                readTreeAndProcessNotificationLock.withLock {
                    val path = getAbsolutePathForProperty(propertyName)
                    val currentData = treeCache.getCurrentData(path)
                    val currentValue = if (currentData == null)
                        null
                    else
                        zkDataToStringOrNull(currentData.data, currentData.toString())
                    accessor(currentValue)
                }
            }
        },
        marshaller = marshaller,
        referenceCleaner = ReferenceCleaner.getInstance()
    )

    private val treeCache: TreeCache

    private val rootPath: String =
        if (zookeeperConfigPath.endsWith('/')) {
            zookeeperConfigPath.dropLast(1)
        } else {
            zookeeperConfigPath
        }

    private val rootPathPrefix = "$rootPath/"

    init {

        if (curatorFramework.state == CuratorFrameworkState.LATENT) {
            curatorFramework.start()
        }
        treeCache = TreeCache(this.curatorFramework, this.rootPath)
        val treeCacheInitialized = CountDownLatch(1)

        treeCache.listenable.addListener(TreeCacheListener { _, treeCacheEvent ->

            logger.trace { "Received TreeCache event: $treeCacheEvent" }

            when (treeCacheEvent.type!!) {
                TreeCacheEvent.Type.NODE_ADDED,
                TreeCacheEvent.Type.NODE_UPDATED -> {
                    try {
                        onZkTreeChanged(
                            treeCacheEvent, zkDataToStringOrNull(
                                treeCacheEvent.data.data,
                                treeCacheEvent.toString()
                            )
                        )
                    } catch (exc: Exception) {
                        logger.error("Zk property updating error for event $treeCacheEvent", exc)
                    }
                }
                TreeCacheEvent.Type.NODE_REMOVED -> {
                    onZkTreeChanged(treeCacheEvent, null)
                }
                TreeCacheEvent.Type.INITIALIZED -> {
                    treeCacheInitialized.countDown()
                }
                TreeCacheEvent.Type.CONNECTION_LOST,
                TreeCacheEvent.Type.CONNECTION_SUSPENDED,
                TreeCacheEvent.Type.CONNECTION_RECONNECTED -> {
                    //do nothing
                }
            }
        })

        treeCache.start()

        if (!treeCacheInitialized.await(initializationTimeout.toMillis(), TimeUnit.MILLISECONDS)) {
            throw ZkDynamicPropertySourceFailedToInitializeOnTime(initializationTimeout)
        }
    }

    /**
     * Works through curator directly to load latest data
     */
    fun readAllProperties(): Map<String, Any> {
        val allProperties = ConcurrentHashMap<String, Any>()

        val exist = curatorFramework.checkExists().forPath(rootPath)
        if (exist != null) {
            val children = curatorFramework.children.forPath(rootPath)
            if (children.isNotEmpty()) {
                val latch = CountDownLatch(children.size)
                for (child in children) {
                    curatorFramework.data.watched().inBackground { _, event ->
                        allProperties[child] = String(event.data, StandardCharsets.UTF_8)
                        latch.countDown()
                    }.forPath(getAbsolutePathForProperty(child))
                }
                if (!latch.await(120, TimeUnit.SECONDS)) {
                    throw TimeoutException("Failed to extract zk properties data")
                }
            }
        }
        return allProperties
    }


    private fun zkDataToStringOrNull(data: ByteArray, logDetails: String): String? =
        try {
            String(data, StandardCharsets.UTF_8)
        } catch (exc: Exception) {
            logger.error("Failed to read string value from zk node. $logDetails", exc)
            null
        }


    private fun onZkTreeChanged(treeCacheEvent: TreeCacheEvent, newValue: String?) {
        val absolutePath = treeCacheEvent.data.path
        if (!absolutePath.startsWith(rootPathPrefix)) {
            return
        }

        val propertyName = getPropertyNameFromAbsolutePath(absolutePath)

        logger.info {
            "Zk property change: type: ${treeCacheEvent.type}, node: $absolutePath. New value is '$newValue'"
        }
        readTreeAndProcessNotificationLock.withLock {
            propertySourcePublisher.notifyAboutPropertyChange(propertyName, newValue)
        }
    }


    private fun getAbsolutePathForProperty(propertyName: String): String {
        Objects.requireNonNull(propertyName)
        return "$rootPath/$propertyName"
    }

    private fun getPropertyNameFromAbsolutePath(absolutePath: String): String {
        return absolutePath.substring(rootPath.length + 1)
    }

    override fun <T : Any?> createSubscription(
        propertyName: String,
        propertyType: Class<T>,
        defaultValue: OptionalDefaultValue<T>
    ): DynamicPropertySource.Subscription<T> {
        return propertySourcePublisher.createSubscription(propertyName, propertyType, defaultValue)
    }

    override fun close() {
        propertySourcePublisher.close()
        treeCache.close()
    }


}

class ZkDynamicPropertySourceFailedToInitializeOnTime(
    timeout: Duration
) : java.lang.Exception(
    "ZkDynamicPropertySource failed to initialize on time. " +
            "Given initialization timeout: $timeout"
)